### PR TITLE
More localization friendly source strings and context commenting

### PR DESCRIPTION
**Pull request by AmelBawa-msft**
# What was done?
- Added support for string localization with placeholders. For example:
```xml
<data name="CommandArgumentDescription" xml:space="preserve">
    <value>Filter results by command</value>
</data>
<data name="InvalidArgumentValueError" xml:space="preserve">
    <value>The value provided for the `{0}` argument is invalid; valid values are: {1}</value>
    <comment>Error message displayed when the user provides an invalid command line argument value. {0} is a placeholder replaced by the argument name. {1} is a placeholder replaced by a list of valid options.</comment>
</data>
```
```cpp
info << String::CommandArgumentDescription;                           // => Filter results by command
info << String::InvalidArgumentValueError("First"_liv, "Second"_liv); // => The value provided for the `First` argument is invalid; valid values are: Second
```
- Applied the localization placeholder change to most use cases in the code. Some remaining use cases (e.g. ShowFlow.cpp) will be handled separately as it contains localized strings with a special rendering style (e.g. text printed in bold, different color, etc...).
- Updated existing unit tests
- Documented strings with placeholders in the `winget.resw` file.

# Test
- Tested locally a randomly selected set of localized strings with placeholders.
- Added unit tests for newly added functions

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2454)

### CODE REVIEW

File: src/AppInstallerCommonCore/Public/winget/ChannelStreams.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#pragma once
```
Comment (AmelBawa-msft) [inline src/AppInstallerCommonCore/Public/winget/ChannelStreams.h:1]: Rename file

File: src/AppInstallerCommonCore/Public/AppInstallerArchitecture.h [hunk 2]
```diff
@@ -23,7 +23,7 @@ namespace AppInstaller::Utility
     Architecture ConvertToArchitectureEnum(const std::string& archStr);
 
     // Converts an Architecture to a string_view
-    LocIndView ToString(Architecture architecture);
+    LocIndView ConvertFromArchitectureEnum(Architecture architecture);
```
Comment (AmelBawa-msft) [inline src/AppInstallerCommonCore/Public/AppInstallerArchitecture.h:26]: Consider using template and keep original name `ToString`

File: src/AppInstallerCommonCore/AppInstallerStrings.cpp [hunk 3]
```diff
@@ -713,4 +705,21 @@ namespace AppInstaller::Utility
...
+            return {};
+        }
+
+        std::ostringstream ssJoin;
+        ssJoin << vector[0];
+        for (size_t i = 1; i < vectorSize; ++i)
+        {
+            ssJoin << separator << vector[i];
+        }
+        return LocIndString{ ssJoin.str() };
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/AppInstallerStrings.cpp:723]: Thanks for the heads up. I just checked those files and noticed that we do have a redundant logic that, in general, might potentially be merged. I created a [task](https://task.ms/40948293) with this suggestion so we can find all redundant implementations and merge them. We might end up creating two different `Join` functions one for `LocIndString` and one `string` (but not mixed) or if possible, have a template implementation.

File: src/AppInstallerCommonCore/Public/AppInstallerStrings.h [hunk 4]
```diff
@@ -230,4 +228,26 @@ namespace AppInstaller::Utility
 
     // Converts the given hexadecimal string into bytes.
     std::vector<uint8_t> ParseFromHexString(const std::string& value, size_t byteCount = 0);
+
+    // Join a string vector using the provided separator.
+    LocIndString Join(LocIndView separator, const std::vector<LocIndString>& vector);
+
+    // Superset of std::to_string supporting string convertibles as input.
+    template <typename T>
+    std::string ToString(T value)
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/AppInstallerStrings.h]: I removed this function shortly before this comment 😄

File: src/AppInstallerCommonCore/Public/winget/ChannelStreams.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/winget/ChannelStreams.h:3]: Oops, added

File: src/AppInstallerCommonCore/Public/winget/ChannelStreams.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        // List of approved types for output, others are potentially not localized.
+        template <typename T>
+        struct IsApprovedForOutput : std::false_type {};
+        WINGET_APPROVED_FOR_OUTPUT_LIST(WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION);
+    }
+
+    template<typename ... T>
+    Utility::LocIndString StringResource::StringId::operator()(T ... args) const
+    {
+        static_assert((Execution::details::IsApprovedForOutput<T>::value && ...), WINGET_ISAPPROVEDFOROUTPUT_ERROR_MESSAGE);
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/winget/ChannelStreams.h]: Fair enough 👍  I reverted this change to match the content from the original implementation in `ChannelStream.h` and kept the comment inline in the `static_assert` for enhanced readability

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 6]
```diff
@@ -676,8 +700,8 @@ They can be configured through the settings file 'winget settings'.</value>
     <value>Shows and performs available upgrades</value>
   </data>
   <data name="Usage" xml:space="preserve">
-    <value>usage</value>
-    <comment>The way to use the software</comment>
+    <value>usage: {0} {1}</value>
+    <comment>Message displayed to provide the user with instructions on how to use a command line command. {0} is a placeholder replaced by the command usage provided by the program.</comment>
```
Comment (JohnMcPMS) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:704]: Missing the difference between `{0}` and `{1}`.

File: src/AppInstallerCommonCore/Public/winget/ChannelStreams.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include "Resources.h"
+
+#define WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION(_t_) template <> struct IsApprovedForOutput<_t_> : std::true_type {};
+#define WINGET_CREATE_ISAPPROVEDFOROUTPUT_STRING_TYPE(_t_) " '" #_t_ "'"
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/winget/ChannelStreams.h]: Reverted the content of this file leaving only one `WINGET_CREATE` macro.

File: src/AppInstallerCommonCore/Public/winget/Resources.h [hunk 8]
```diff
@@ -55,5 +62,36 @@ namespace AppInstaller
...
+        {
+            ResourceOpenException(const winrt::hresult_error& hre);
+
+            const char* what() const noexcept override { return m_message.c_str(); }
+
+        private:
+            std::string m_message;
+        };
+
+        // TODO https://task.ms/40934960 Consider renaming or removing LocString
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/winget/Resources.h:76]: @lechacon, `LocString` was responsible for taking `StringId` and resolve it from the `.resw` file. Now that we can produce `LocIndString` directly using the added `StringId::operator()` operator, we can directly store resolved strings in `LocIndString` without using `LocString`, and for `StringId` instances (the case of localized strings with no placeholders) we can add a constructor in `LocIndString` which simply calls the `()` with no parameters (which what `LocString` does in this PR).

@JohnMcPMS, removing `LocString` struct will result in a lot of changes across many files which I wanted to avoid in this PR. But I created a task to better evaluate this change along with any unexpected compilations issues.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 9]
```diff
@@ -118,44 +118,52 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="AdjoinedNotFlagError" xml:space="preserve">
-    <value>Adjoined alias is not a flag</value>
+    <value>Adjoined alias is not a flag: '{0}'</value>
+    <comment>Error message displayed when the user provides an adjoined alias that is not a flag argument. {0} is a placeholder replaced by the user input argument (e.g. '-ab').</comment>
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:122]: To add an example:
```ps
winget install vscode -hm
  -h silent
  -m manifest (expects input, hence not a flag)
```

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 10]
```diff
@@ -118,44 +118,52 @@
...
   </data>
   <data name="AvailableCommands" xml:space="preserve">
     <value>The following commands are available:</value>
-    <comment>Commands the tool supports</comment>
+    <comment>Message displayed to inform the user about the available command line command arguments.</comment>
   </data>
   <data name="AvailableHeader" xml:space="preserve">
     <value>Available</value>
     <comment>As in "a new version is available to update to".</comment>
+    <comment>Label displayed when a new application package is available.</comment>
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:143]: Unlikely since Visual studio only allow one comment (and extra comments will be removed on save). But we can have multiple lines comments if we want to add separate information.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 11]
```diff
@@ -118,44 +118,52 @@
...
   </data>
   <data name="AvailableOptions" xml:space="preserve">
     <value>The following options are available:</value>
+    <comment>Message displayed to inform the user about the available options.</comment>
   </data>
   <data name="AvailableSubcommands" xml:space="preserve">
     <value>The following sub-commands are available:</value>
-    <comment>Nested commands that can be run in context of the selected command</comment>
+    <comment>Message displayed to inform the user about the available nested command line command arguments that runs in context of the selected command.</comment>
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: I added a prefix for all new comments primarily to distinguish between info, error and other type of messages (or title, as you suggested in a previous comment), giving more context during the translation process. If you have suggestion for an alternative prefix, let me know.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 12]
```diff
@@ -118,44 +118,52 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:118]: Yes, I have a work item tracking a list of localization related things we can improve. Sorting `resw` is one of the items😃

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 13]
```diff
@@ -1238,8 +1286,8 @@ Please specify one of them using the `--source` option to proceed.</value>
     <comment>{Locked="https://github.com/microsoft/winget-cli"}</comment>
   </data>
   <data name="WindowsPackageManager" xml:space="preserve">
-    <value>Windows Package Manager</value>
-    <comment>The product name.</comment>
+    <value>Windows Package Manager v{0}</value>
```
[the remark this answers is not in the document]
Comment (denelon) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1287]: I'm tempted to drop the v so it's easier to transform into a PowerShell version and avoid the need to drop the character.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 9 above)
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:122]: Thanks, I didn't know that.
Is that a commonly understood term that the loc team knows and can translate? Otherwise we should explain it.

File: src/AppInstallerCommonCore/Public/AppInstallerStrings.h [hunk 14]
```diff
@@ -230,4 +228,26 @@ namespace AppInstaller::Utility
...
+
+    // Format an input string by replacing placeholders {index} with provided values at corresponding indices.
+    // Note: After upgrading to C++20, this function should be deprecated in favor of std::format.
+    template <typename ... T>
+    std::string Format(std::string inputStr, T ... args)
+    {
+        int index = 0;
+        (FindAndReplace(inputStr, "{" + std::to_string(index++) + "}", ToString(args)), ...);
+        return inputStr;
+    }
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/AppInstallerStrings.h]: We only use the basic format functionality, which is why I went with a one-liner variadic template function. But if we decided we want more complex format, we can implement a simple algorithm that iterates and replaces placeholders, or even upgrade to C++20 to benefit from `std::format` OOB 😄

File: src/AppInstallerCommonCore/Public/AppInstallerStrings.h [hunk 15]
```diff
@@ -230,4 +228,26 @@ namespace AppInstaller::Utility
...
+
+    // Format an input string by replacing placeholders {index} with provided values at corresponding indices.
+    // Note: After upgrading to C++20, this function should be deprecated in favor of std::format.
+    template <typename ... T>
+    std::string Format(std::string inputStr, T ... args)
+    {
+        int index = 0;
+        (FindAndReplace(inputStr, "{" + std::to_string(index++) + "}", ToString(args)), ...);
+        return inputStr;
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/AppInstallerStrings.h]: We can scan the string after the one liner, but if we do, I would consider implementing the scan and replace placeholders approach instead since we can do all checks in one run.

File: src/AppInstallerCLICore/Workflows/UninstallFlow.cpp [hunk 16]
```diff
@@ -252,17 +252,22 @@ namespace AppInstaller::CLI::Workflow
...
-                context.Reporter.Error() << Resource::String::UninstallFailedWithCode << ' ' << GetUserPresentableMessage(uninstallResult) << std::endl;
+                context.Reporter.Error()
+                    << Resource::String::UninstallFailedWithCode(Utility::LocIndView{ GetUserPresentableMessage(uninstallResult) })
+                    << std::endl;
             }
             else
             {
-                context.Reporter.Error() << Resource::String::UninstallFailedWithCode << ' ' << uninstallResult << std::endl;
+                context.Reporter.Error()
+                    << Resource::String::UninstallFailedWithCode(Utility::LocIndView{ std::to_string(uninstallResult) })
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/UninstallFlow.cpp:262]: static assert will not allow `DWORD` type, but we can extend the approved for output types to include it (and possibly other numeric types)

File: src/AppInstallerCLICore/Core.cpp [hunk 17]
```diff
@@ -135,8 +135,9 @@ namespace AppInstaller::CLI
         {
             // Report any action blocked by Group Policy.
             auto policy = Settings::TogglePolicy::GetPolicy(e.Policy());
+            auto policyNameId = policy.PolicyName();
             AICLI_LOG(CLI, Error, << "Operation blocked by Group Policy: " << policy.RegValueName());
-            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy << " : "_liv << policy.PolicyName() << std::endl;
+            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy(policyNameId()) << std::endl;
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Core.cpp]: Updated. Passing `StringId` should be sufficient now + added test case for this 👍

File: src/AppInstallerCLICore/Workflows/DependencyNodeProcessor.cpp [hunk 18]
```diff
@@ -52,14 +53,14 @@ namespace AppInstaller::CLI::Workflow
         
         if (!m_nodePackageLatestVersion)
         {
-            error << Resource::String::DependenciesFlowPackageVersionNotFound << " " << Utility::Normalize(packageId);
+            error << Resource::String::DependenciesFlowPackageVersionNotFound(Utility::LocIndView{ Utility::Normalize(packageId) });
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/DependencyNodeProcessor.cpp:56]: Added this suggestion for consideration in a separate PR.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 13 above)
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1287]: Resolving thread, please re-activate if we should consider adjusting the corresponding label values in the resource file.

File: src/AppInstallerCLICore/Workflows/SourceFlow.cpp [hunk 19]
```diff
@@ -258,7 +258,7 @@ namespace AppInstaller::CLI::Workflow
 
         for (const auto& source : sources)
         {
-            context.Reporter.Info() << Resource::String::SourceResetOne << ' ' << source.Name << "..."_liv;
+            context.Reporter.Info() << Resource::String::SourceResetOne(Utility::LocIndView{ source.Name });
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/SourceFlow.cpp:261]: Good point, the reason I did not change certain field/function types (including this suggestion) is because those updates will trigger much more changes in many other files which I am trying to avoid doing in a single PR. That said, I am tracking internally suggestions and opportunities for enhancements which would be considered in separate PRs.

File: src/AppInstallerCLICore/Resources.h [hunk 20]
```diff
@@ -128,7 +129,6 @@ namespace AppInstaller::CLI::Resource
         WINGET_DEFINE_RESOURCE_STRINGID(InstallationDisclaimer1);
         WINGET_DEFINE_RESOURCE_STRINGID(InstallationDisclaimer2);
         WINGET_DEFINE_RESOURCE_STRINGID(InstallationDisclaimerMSStore);
-        WINGET_DEFINE_RESOURCE_STRINGID(InstallationRequiresHigherWindows);
```
Comment (florelis) [inline src/AppInstallerCLICore/Resources.h:131]: I'm curious how this string ended up unused

File: src/AppInstallerCLICore/Command.cpp [hunk 21]
```diff
@@ -114,7 +80,7 @@ namespace AppInstaller::CLI
         }
 
         // Output the command preamble and command chain
-        infoOut << Resource::String::Usage << ": winget"_liv << Utility::LocIndView{ commandChain };
+        infoOut << Resource::String::Usage("winget"_lis, Utility::LocIndView{ commandChain });
```
Comment (florelis) [inline src/AppInstallerCLICore/Command.cpp:83]: Why did `_liv` become `_lis`?

File: src/AppInstallerCLICore/Command.cpp [hunk 22]
```diff
@@ -279,10 +245,10 @@ namespace AppInstaller::CLI
         }
 
         // Finally, the link to the documentation pages
-        std::string helpLink = HelpLink();
+        auto helpLink = Utility::LocIndString{ HelpLink() };
```
Comment (florelis) [inline src/AppInstallerCLICore/Command.cpp:248]: Nit: Doesn't seem too hard to change this one to be loc ind to begin with

File: src/AppInstallerCLITests/WorkFlow.cpp [hunk 23]
```diff
@@ -1190,7 +1191,7 @@ TEST_CASE("InstallFlow_Zip_BadRelativePath", "[InstallFlow][workflow]")
 
     // Verify Installer was not called
     REQUIRE(!std::filesystem::exists(installResultPath.GetPath()));
-    REQUIRE(installOutput.str().find(Resource::LocString(Resource::String::NestedInstallerNotFound).get()) != std::string::npos);
+    REQUIRE(installOutput.str().find(Resource::LocString(Resource::String::NestedInstallerNotFound(""_liv)).get()) != std::string::npos);
```
Comment (florelis) [inline src/AppInstallerCLITests/WorkFlow.cpp:1194]: Why `""_liv`?

File: src/AppInstallerCLITests/WorkFlow.cpp [hunk 24]
```diff
@@ -2445,7 +2446,7 @@ TEST_CASE("UpdateFlow_UpdateExeSpecificVersionNotFound", "[UpdateFlow][workflow]
 
     // Verify Installer is not called.
     REQUIRE(!std::filesystem::exists(updateResultPath.GetPath()));
-    REQUIRE(updateOutput.str().find(Resource::LocString(Resource::String::GetManifestResultVersionNotFound).get()) != std::string::npos);
+    REQUIRE(updateOutput.str().find(Resource::LocString(Resource::String::GetManifestResultVersionNotFound("1.2.3.4"_liv)).get()) != std::string::npos);
```
Comment (florelis) [inline src/AppInstallerCLITests/WorkFlow.cpp:2449]: I like how these tests became more specific :)

File: src/AppInstallerCLITests/Completion.cpp [hunk 25]
```diff
@@ -233,7 +233,7 @@ struct CompletionTestCommand : public Command
 
     using Command::Complete;
 
-    void Complete(Execution::Context& context, Execution::Args::Type valueType) const override
+    void Complete(CLI::Execution::Context& context, CLI::Execution::Args::Type valueType) const override
```
Comment (florelis) [inline src/AppInstallerCLITests/Completion.cpp:236]: Why did you need to do this?

**Review by florelis — APPROVED** [review]

File: src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp [hunk 26]
```diff
@@ -148,9 +148,13 @@ namespace AppInstaller::CLI::Workflow
         }
         else
         {
-            context.Reporter.Info() << Resource::String::MSStoreInstallOrUpdateFailed << ' ' << WINGET_OSTREAM_FORMAT_HRESULT(errorCode) << std::endl;
+            std::ostringstream ssError;
+            ssError << WINGET_OSTREAM_FORMAT_HRESULT(errorCode);
+            auto errorCodeString = Utility::LocIndString{ ssError.str() };
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp:153]: These 3 lines can be in its own method that returns a LocIndString so that can be reused in other places that needs to print HResult

File: src/AppInstallerCommonCore/Resources.cpp [hunk 27]
```diff
@@ -4,55 +4,121 @@
...
+        }
+
+        std::pair<const BYTE*, size_t> GetResourceAsBytes(PCWSTR resourceName, PCWSTR resourceType)
+        {
+            auto resourceData = GetResourceData(resourceName, resourceType);
+            return std::make_pair(reinterpret_cast<BYTE*>(resourceData.first), resourceData.second);
+        }
+
+        // Utility class to load resources
+        class Loader
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Resources.cpp:64]: >class Loader [](http://example.com/codeflow?start=8&length=12)

nit: we use struct across the project and below public: could be removed.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 28]
```diff
@@ -118,44 +118,51 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="AdjoinedNotFlagError" xml:space="preserve">
-    <value>Adjoined alias is not a flag</value>
+    <value>Adjoined alias is not a flag: '{0}'</value>
```
Comment (yao-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:121]: >{0} [](http://example.com/codeflow?start=42&length=3)

Use {Locked="{0}","{1}"} in the comment across all the changes that added placeholders across this file to prevent human errors

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 29]
```diff
@@ -420,7 +434,6 @@ They can be configured through the settings file 'winget settings'.</value>
   </data>
   <data name="Package" xml:space="preserve">
     <value>Package</value>
-    <comment>A software package</comment>
```
Comment (yao-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:423]: ><comment>A software package</comment> [](http://example.com/codeflow?start=4&length=37)

Keep this comment to avoid translator translating to other "package"

File: src/AppInstallerCLICore/ChannelStreams.h [hunk 30]
```diff
@@ -11,36 +11,6 @@
...
-        WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION(Utility::LocIndView);
-        WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION(Utility::LocIndString);
-        // Normalized strings come from user data and should therefore already by localized
-        // by how they are chosen (or there is no localized version).
-        WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION(Utility::NormalizedString);
-    }
-
     // The base stream for all channels.
     struct BaseStream
     {
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/ChannelStreams.h:16]: Good point, I will update the TODO message here. Unfortunately, uncommenting this `static_assert` will require a completely new set of changes, but it's certainly doable (maybe in several smaller PRs?). That said, with the new localization syntax introduced in this PR, this rule will be enforced on all localized strings placeholders, which is a good step in solving part of the problem, and ultimately it will limit the number of changes required to uncomment this in the future.

File: src/AppInstallerCLICore/Command.cpp (hunk 22 above)
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Command.cpp:248]: Correct, this one is simple however it does require changes in around 30+ files. I am trying to avoid making too many changes in a single PR to help maintain a controllable scope. I have this suggestion tracked separately as a follow up 😄

File: src/AppInstallerCommonCore/Public/winget/Resources.h [hunk 31]
```diff
@@ -20,6 +21,16 @@ namespace AppInstaller
         struct StringId : public std::wstring_view
         {
             explicit constexpr StringId(std::wstring_view id) : std::wstring_view(id) {}
+
+            // Sets the placeholder values in the resolved string id.
+            template<typename ...T>
+            Utility::LocIndString operator()(T ... args) const;
+
+            friend std::ostream& operator<<(std::ostream& out, StringId si);
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCommonCore/Public/winget/Resources.h]: `friend` is not needed here, removed keyword and moved function directly under the same namespace.
`Resolve` is private to ensure we always use the new syntax (`operator()`) which calls the function internally (if the resolved string id expects arguments, the new syntax will `static_assert` the types to prevent user errors).

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp [hunk 32]
```diff
@@ -259,15 +259,10 @@ namespace AppInstaller::CLI::Workflow
...
         {
             auto policy = Settings::TogglePolicy::GetPolicy(e.Policy());
-            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy << ": "_liv << policy.PolicyName() << std::endl;
+            auto policyNameId = policy.PolicyName();
+            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy(policyNameId()) << std::endl;
             return APPINSTALLER_CLI_ERROR_BLOCKED_BY_POLICY;
         }
-        catch (const Resource::ResourceOpenException& e)
-        {
-            Logging::Telemetry().LogException(Logging::FailureTypeEnum::ResourceOpen, e.what());
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:267]: We can since it's not used anymore, but keeping it helps us remember to not reuse the same code for another failure type in the future.

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp [hunk 33]
```diff
@@ -579,7 +574,7 @@ namespace AppInstaller::CLI::Workflow
                 auto warn = context.Reporter.Warn();
                 for (const auto& failure : searchResult.Failures)
                 {
-                    warn << Resource::String::SearchFailureWarning << ' ' << failure.SourceName << std::endl;
+                    warn << Resource::String::SearchFailureWarning(Utility::LocIndView{ failure.SourceName }) << std::endl;
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:577]: You called it 😆 I tried changing it but that triggered many compiler errors requiring a lot of changes. Moving forward, we can incrementally make those enhancements as needed, or have separate PRs if the changes are too big.

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp [hunk 34]
```diff
@@ -259,15 +259,10 @@ namespace AppInstaller::CLI::Workflow
         catch (const Settings::GroupPolicyException& e)
         {
             auto policy = Settings::TogglePolicy::GetPolicy(e.Policy());
-            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy << ": "_liv << policy.PolicyName() << std::endl;
+            auto policyNameId = policy.PolicyName();
+            context.Reporter.Error() << Resource::String::DisabledByGroupPolicy(policyNameId()) << std::endl;
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Workflows/WorkflowBase.cpp]: Updated. The `()` are not needed anymore (optinoal) since the implemented operator `()` overload will take care of resolving it.

File: src/AppInstallerCLICore/Command.cpp (hunk 21 above)
Comment (AmelBawa-msft) [reply src/AppInstallerCLICore/Command.cpp]: Typo, fixed👍

File: src/AppInstallerCLITests/WorkFlow.cpp (hunk 23 above)
Comment (AmelBawa-msft) [reply src/AppInstallerCLITests/WorkFlow.cpp]: Updated to use the expected relative path for more accurate string loc comparison.

File: src/AppInstallerCLITests/Completion.cpp (hunk 25 above)
Comment (AmelBawa-msft) [reply src/AppInstallerCLITests/Completion.cpp]: I believe that was needed when I was moving code from Core to Common (IIRC ChannelStream related code). It's not needed anymore, good catch!

**Review by yao-msft — APPROVED** [review]
:shipit:
